File: daily/Arrays/777.Swap-adjacent-in-lr-string.go
package daily

func canTransform(start string, result string) bool {
	left, right := 0, 0
	n := len(start)
	for left < len(start) && right < len(result) {
		for left < len(start) && start[left] == 'X' {
			left++
		}
		for right < len(result) && result[right] == 'X' {
			right++
		}

		if left == n || right == n {
			break
		}

		if start[left] != result[right] {
			return false
		}
		if start[left] == 'L' && left < right {
			return false
		}
		if start[left] == 'R' && left > right {
			return false
		}

		left++
		right++
	}

	for left < n {
		if start[left] != 'X' {
			return false
		}
		left++
	}
	for right < n {
		if result[right] != 'X' {
			return false
		}
		right++
	}

	return true
}

File: daily/Arrays/777.Swap-adjacent-in-lr-string.py
class Solution:
    def canTransform(self, start: str, end: str) -> bool:
        i = j = 0
        n = len(start)

        while i < n and j < n:
            # Skip 'X' in both strings
            while i < n and start[i] == 'X':
                i += 1
            while j < n and end[j] == 'X':
                j += 1

            if i == n or j == n:
                break

            if start[i] != end[j]:
                return False

            # 'L' can only move left: its index cannot increase
            if start[i] == 'L' and i < j:
                return False
            # 'R' can only move right: its index cannot decrease
            if start[i] == 'R' and i > j:
                return False

            i += 1
            j += 1

        # Remaining characters must be 'X' in both strings
        while i < n:
            if start[i] != 'X':
                return False
            i += 1
        while j < n:
            if end[j] != 'X':
                return False
            j += 1

        return True

